### PR TITLE
Galaxy

### PR DESCRIPTION
**Pull request by dsavchenko**
- [x] sometimes creates duplicated PRs

### CODE REVIEW

**Comment by dsavchenko** [comment]
Example of the automatic PR https://github.com/esg-epfl-apc/tools-astro/pull/46

What's missing now, is *.shed.yml* file generation. But here, need to decide where this info (at least description and long_description are needed) comes from. Does workflow creator need to create a file with some fields defined, or just provide full *.shed.yml*?

File: odabot/cli.py [hunk 1]
```diff
@@ -129,16 +134,19 @@ def cli(obj, debug, settings):
     )
 
     obj['debug'] = debug
-    
-    logger.info("components: %s", obj['settings'].components)
+    try:
+        logger.info("components: %s", obj['settings'].components)
+    except AttributeError:
+        pass
```
Comment (volodymyrss) [inline odabot/cli.py:140]: could you add some log here?

Comment (dsavchenko) [reply odabot/cli.py]: Better to move it to the place where it shouldn't fail

File: requirements_galaxy.txt [hunk 2]
```diff
@@ -0,0 +1,2 @@
+nb2workflow @ git+https://github.com/oda-hub/nb2workflow@galaxy
+oda_api @ git+https://github.com/oda-hub/oda_api@ontology-helper
```
Comment (volodymyrss) [inline requirements_galaxy.txt:2]: this is merged already

**Comment by volodymyrss** [comment]
branches should be probably created right in org repo, not in your own. Else secrets are not used (and they should not be)

**Comment by dsavchenko** [comment]
> branches should be probably created right in org repo

Yes, it is a configurable parameter

**Review by volodymyrss — DISMISSED** [review]

**Comment by bgruening** [comment]
I don't know :-( can you explain your use-case? Maybe we can brainstorm how tomsolve it differently.

**Comment by dsavchenko** [comment]
> I don't know :-( can you explain your use-case? Maybe we can brainstorm how tomsolve it differently.

This is for the bot to decide the version of the next autogenerated tool release. I thought of the toolshed to be a single source of truth for the latest version already available. But if this is complicated, I can rely on the version of the tool in the main branch of our repo https://github.com/esg-epfl-apc/tools-astro/ assuming it's the only one with CI configured to push into the toolshed

**Comment by bgruening** [comment]
I see. We use GitHub. We follow those recommendations for numbering https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html and then let our bot increase the number.

**Comment by volodymyrss** [comment]
@dsavchenko it seems ready and approved. Maybe let's merge it and make further changes with new PRs as needed? Or something essential is missing?

**Comment by dsavchenko** [comment]
> @dsavchenko it seems ready and approved. Maybe let's merge it and make further changes with new PRs as needed? Or something essential is missing?

I wanted to merge it when the first generated tool is ready to be pushed to the toolshed (we are very close to this).
I will re-request review

**Comment by dsavchenko** [comment]
When https://github.com/oda-hub/nb2workflow/pull/102 is merged, the nb2workflow branch needs to be changed in `Dockerfile_galaxy` and this PR will also be ready.  I will add CI/CD to build galaxy-related container separately in the future.

The first autogenerated tool PR https://github.com/esg-epfl-apc/tools-astro/pull/82 looks good now.

File: odabot/cli.py [hunk 3]
```diff
@@ -651,7 +661,267 @@ def verify_workflows(obj):
...
+                        
+                        logger.info('last_commit_created_at %s saved_last_commit_created_at %s', last_commit_created_at, saved_last_commit_created_at )
+
+                        if last_commit_created_at == saved_last_commit_created_at and not force:
+                            logger.info("no need to deploy this tool")
+                        else:
+                            wf_repo_dir = os.path.join(repo_cache_dir, project['path'])
+                            git_clone_or_update(wf_repo_dir, project['http_url_to_repo'])
+                        
+                            req_file = os.path.join(wf_repo_dir, 'requirements.txt') if os.path.isfile(os.path.join(wf_repo_dir, 'requirements.txt')) else None
```
Comment (volodymyrss) [inline odabot/cli.py:790]: this is a very long line. And the indent is very deep too. Could it be taken into a function?

File: odabot/cli.py [hunk 4]
```diff
@@ -4,32 +4,39 @@
...
+try:
+    from nb2workflow.galaxy import to_galaxy    
+    import frontmatter
+except ImportError:
+    logger.warning('Galaxy dependencies not loaded')
 
-# `envvar_prefix` = export envvars with `export DYNACONF_FOO=bar`.
-# `settings_files` = Load this files in the order.
 
 renkuapi = "https://gitlab.renkulab.io/api/v4/"
```
Comment (volodymyrss) [inline odabot/cli.py:41]: not added here, but shouldn't this be in config? Especially if we want to use different gitlabs.

**Review by volodymyrss — APPROVED** [review]

File: odabot/cli.py (hunk 4 above)
Comment (dsavchenko) [reply odabot/cli.py:41]: will make a separate small pr

**Pull request merged**